// axios-typescript/dist/lib/core/dispatchRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var xhr_1 = require("../xhr");
var url_1 = require("../helpers/url");
var headers_1 = require("../helpers/headers");
var transform_1 = require("./transform");
function dispatchRequest(config) {
    throwIfCancellationRequest(config);
    processConfig(config);
    return xhr_1.default(config).then(function (res) {
        return transformResponsetData(res);
    });
}
exports.default = dispatchRequest;
function processConfig(config) {
    config.url = transformUrl(config);
    // config.headers = transformHeaders(config)
    config.data = transform_1.default(config.data, config.headers, config.transformRequest);
    config.headers = headers_1.flattenHeaders(config.headers, config.method);
}
function transformUrl(config) {
    var url = config.url, params = config.params, paramsSerializer = config.paramsSerializer, baseURL = config.baseURL;
    if (baseURL && !url_1.isAbsolutURL(url)) {
        url = url_1.combineURL(baseURL, url);
    }
    return url_1.buildURL(url, params, paramsSerializer);
}
exports.transformUrl = transformUrl;
// function transformRequestData(config: AxiosRequestConfig): any {
//   return transformRequest(config.data)
// }
// function transformHeaders(config: AxiosRequestConfig): any {
//   const { headers = {}, data } = config
//   return processHeaders(headers, data)
// }
function transformResponsetData(res) {
    res.data = transform_1.default(res.data, res.headers, res.headers.transformResponse);
    return res;
}
function throwIfCancellationRequest(config) {
    if (config.cancelToken) {
        config.cancelToken.throwIfRequested();
    }
}
